cranelift-wasm/src/state/module_state.rs

```diff
@@ -24,4 +31,361 @@ impl ModuleTranslationState {
             wasm_types: PrimaryMap::new(),
         }
     }
+
+    /// Translate a binary WebAssembly function.
+    ///
+    /// The `code` slice contains the binary WebAssembly *function code* as it appears in the code
+    /// section of a WebAssembly module, not including the initial size of the function code. The
+    /// slice is expected to contain two parts:
+    ///
+    /// - The declaration of *locals*, and
+    /// - The function *body* as an expression.
+    ///
+    /// See [the WebAssembly specification][wasm].
+    ///
+    /// [wasm]: https://webassembly.github.io/spec/core/binary/modules.html#code-section
+    ///
+    /// The Cranelift IR function `func` should be completely empty except for the `func.signature`
+    /// and `func.name` fields. The signature may contain special-purpose arguments which are not
+    /// regarded as WebAssembly local variables. Any signature arguments marked as
+    /// `ArgumentPurpose::Normal` are made accessible as WebAssembly local variables.
+    ///
+    pub fn translate_func<FE: FuncEnvironment + ?Sized>(
+        &self,
+        code: &[u8],
+        code_offset: usize,
+        func: &mut ir::Function,
+        environ: &mut FE,
+    ) -> WasmResult<()> {
+        self.translate_func_with_reader(
+            BinaryReader::new_with_offset(code, code_offset),
+            func,
+            environ,
+        )
+    }
+
+    /// Translate a binary WebAssembly function with a `BinaryReader` rather
+    /// than `[u8]` slice.
+    pub fn translate_func_with_reader<FE: FuncEnvironment + ?Sized>(
+        &self,
+        mut reader: BinaryReader,
+        func: &mut ir::Function,
+        environ: &mut FE,
+    ) -> WasmResult<()> {
+        let _tt = timing::wasm_translate_function();
+        info!(
+            "translate({} bytes, {}{})",
+            reader.bytes_remaining(),
+            func.name,
+            func.signature
+        );
+        debug_assert_eq!(func.dfg.num_ebbs(), 0, "Function must be empty");
+        debug_assert_eq!(func.dfg.num_insts(), 0, "Function must be empty");
+
+        let mut func_ctx = FunctionBuilderContext::new();
```

Comment:
This will raise the number of mallocs by a large number. This context structure only exists for the purpose of sharing memory across function compilations. Can you store it in the `ModuleTranslationState` instead, and pass it by mut ref below?